### PR TITLE
get instances

### PR DESCRIPTION
**Pull request by ArtificialOwl**
light version of #43, for php 7.4

### CODE REVIEW

File: server/lib/InstanceManager.php [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+
+	public function getAllFromConfig(): array {
+		return $this->instances;
+	}
+
+	/**
+	 * sync the instances from the users table
+	 */
+	public function syncInstances(): void {
```
Comment (come-nc) [inline server/lib/InstanceManager.php:124]: Can you add comments to this one? It’s not clear what it does.

File: server/lib/InstanceManager.php [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+		foreach ($instances as $instance) {
+			$this->insert($instance);
+		}
+
+		$this->removeDeprecatedInstances($instances);
+	}
+
+
+	/**
+	 * @param string|null $instance
```
Comment (come-nc) [inline server/lib/InstanceManager.php:149]:
```suggestion
	 * @param string $instance
```

File: server/lib/InstanceManager.php [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+		if ($removeUsers) {
+			$this->removeUsers($instance);
+		}
+	}
+
+
+	/**
+	 * @param string $instance
+	 */
+	private function removeUsers(string $instance) {
```
Comment (come-nc) [inline server/lib/InstanceManager.php:167]:
```suggestion
	private function removeUsers(string $instance): void {
```

File: server/lib/InstanceManager.php [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+		$stmt->closeCursor();
+		$this->removingEmptyInstance($instance);
+	}
+
+
+	/**
+	 * @param int $userId
+	 */
+	private function removeUser(int $userId) {
```
Comment (come-nc) [inline server/lib/InstanceManager.php:185]:
```suggestion
	private function removeUser(int $userId): void {
```

File: server/lib/InstanceManager.php [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+		$instance = substr($cloudId, $pos + 1);
+
+		$this->removingEmptyInstance($instance);
+	}
+
+
+	/**
+	 * @param string $instance
+	 */
+	private function removingEmptyInstance(string $instance) {
```
Comment (come-nc) [inline server/lib/InstanceManager.php:234]:
```suggestion
	private function removingEmptyInstance(string $instance): void {
```

File: server/lib/SignatureHandler.php [hunk 6]
```diff
@@ -97,4 +98,33 @@ private function splitCloudId(string $cloudId): array {
...
+	 * @param Request $request
+	 *
+	 * @throws SignedRequestException
+	 */
+	public function verifyRequest(Request $request): string {
+		$body = json_decode($request->getBody(), true);
+		if ($body === null || !isset($body['message']) || !isset($body['message']['data'])
+			|| !isset($body['message']['data']['federationId'])
+			|| !isset($body['signature'])
+			|| !isset($body['message']['timestamp'])) {
```
Comment (come-nc) [inline server/lib/SignatureHandler.php:112]:
```suggestion
		if ($body === null ||
			|| !isset($body['message']['data']['federationId'])
			|| !isset($body['signature'])
			|| !isset($body['message']['timestamp'])) {
```
isset does not need to test each intermediate array

File: server/lib/SignatureHandler.php [hunk 7]
```diff
@@ -97,4 +98,33 @@ private function splitCloudId(string $cloudId): array {
...
+
+		try {
+			$verified = $this->verify($cloudId, $body['message'], $body['signature']);
+			if ($verified) {
+				list(, $host) = $this->splitCloudId($body['message']['data']['federationId']);
+
+				return $host;
+			}
+		} catch (\Exception $e) {
+		}
```
Comment (come-nc) [inline server/lib/SignatureHandler.php:126]: Maybe throw the SignedRequestException with $e as previous in here? Might help debugging later.

File: server/lib/SignatureHandler.php [hunk 8]
```diff
@@ -97,4 +98,33 @@ private function splitCloudId(string $cloudId): array {
...
+			|| !isset($body['message']['timestamp'])) {
+			throw new SignedRequestException();
+		}
+
+		$cloudId = $body['message']['data']['federationId'];
+
+		try {
+			$verified = $this->verify($cloudId, $body['message'], $body['signature']);
+			if ($verified) {
+				list(, $host) = $this->splitCloudId($body['message']['data']['federationId']);
```
Comment (come-nc) [inline server/lib/SignatureHandler.php:121]:
```suggestion
				[, $host] = $this->splitCloudId($body['message']['data']['federationId']);
```
For consistency with Nextcloud code

File: server/lib/InstanceManager.php [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+		if ($data === false) {
+			$time = time();
+			$insert = $this->db->prepare(
+				'INSERT INTO instances (instance, timestamp) VALUES (:instance, FROM_UNIXTIME(:timestamp))'
+			);
+			$insert->bindParam(':instance', $instance, PDO::PARAM_STR);
+			$insert->bindParam(':timestamp', $time, PDO::PARAM_INT);
+
+			$insert->execute();
+		}
```
[the remark this answers is not in the document]
Comment (ArtificialOwl) [reply server/lib/InstanceManager.php:50]: `instance` is unique. exception on the `qb->execute()` is ignored

[the remark this answers is not in the document]
Comment (come-nc) [reply server/lib/InstanceManager.php]: But then why is there a select first to see if it exists?

File: server/lib/SignatureHandler.php [hunk 10]
```diff
@@ -97,4 +98,34 @@ private function splitCloudId(string $cloudId): array {
...
+
+		try {
+			$verified = $this->verify($cloudId, $body['message'], $body['signature']);
+			if ($verified) {
+				[, $host] = $this->splitCloudId($body['message']['data']['federationId']);
+
+				return $host;
+			}
+		} catch (\Exception $e) {
+			throw new SignedRequestException($e->getMessage());
```
Comment (come-nc) [inline server/lib/SignatureHandler.php:126]:
```suggestion
			throw new SignedRequestException($e->getMessage(), 0, $e);
```

File: server/lib/InstanceManager.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (ArtificialOwl) [reply server/lib/InstanceManager.php]: select might be lighter, resourcewise. and 99.99% of the time the instance will already exists

**Pull request merged**